src/Microsoft.Extensions.Configuration/ConfigurationRoot.cs

```diff
@@ -24,6 +24,12 @@ public ConfigurationRoot(IList<IConfigurationProvider> providers)
             _providers = providers;
         }
 
+        public string Key => string.Empty;
```

Comment:
Per user request in #302 empty keys are valid keys and so it is valid to call `root.GetSection(string.Empty)` AKA `GetSubsection(string.Empty)` to get a __child__ `IConfiguration`. Now, that call would return an `IConfiguration` for which `Key` and `Path` would also be `string.Emtpty`. At that point what is the difference between the root and the child?

My point is that this seems broken and probably assumptions made in other places are broken. 

- One possible solution (that I am not really advocating for!) is to change the implementation of `ConfigurationRoot.GetSection(key)` to do the same thing that `ConfigurationSection.GetSection(key)` which is would concatenate `Path + Constants.KeyDelimiter + key`. That would imply that all the paths would start with and empty key, e.g. ":x:y:z". 
- Another possible solution is to say that the key of the root are actually `null` and not `string.Empty`. At that point you would have exchanged the problem of having to try cast/is check `IConfguration` instances to see if they are actually `IConfigurationRoot` or `IConfigurationSection` for having to guard against nulls, at which point I would argue you have gotten very little value for this change.